src/makeddr/makeddr.1
NAME:
  makeddr  --  creates a ddr file from the command line.

SYNOPSIS:
  makeddr [<projection> <specific parameters> <ULn> <ULe> <LRn> <LRe> <pdist>]
	  [-log <file>] [-d <datum code] <filename> <nl> <ns> <type>
	  
DESCRIPTION: 
	Create a LAS 6.0 DDR (Data Descriptor Record) from the command line.
  This is only useful when you don't have a DDR already (which is actually
  pretty rare, but when you need one it's tough to get along without it).
  This might happen if you're importing data from an source that doesn't use
  LAS (the Land Analysis System).

  The required inputes are:
    filename	The created DDR is called "filename.ddr".
    nl		Describes an image with a height of nl lines.
    ns		Describes an image with a width of ns samples.
    type	The data types are:

	data type	Description
	----------------------------------------------------------------
	byte		unsigned 1 byte integer data
	short		signed 2 byte integer data
	long		signed 4 byte integer data
	float		IEEE 4 byte single-precision floating-point data
	double		IEEE 8 byte double-precision floating-point data


OPTIONS:
  -log	Option to have output written to a log <file>.

  -d	Option to specify a <datum code>
  	  The datum code is a predefined number describing the eccentricity
	  of the earth. Here is a table:
Code       Name 	     Equitorial Radius	 Flattening	Polar Radius  Eccentricity^2
--------------------------------------------------------------------------------------------
  0   CLARKE 1866		6378206.4	294.9786982	6356583.8	0.006768658
  1   CLARKE 1880		6378249.145	293.465		6356514.87	0.006803511
  2   BESSEL			6377397.155	299.1528128	6356078.963	0.006674372
  3   INTERNATIONAL 1967	6378157.5	298.25		6356772.228	0.006694542
  4   INTERNATIONAL 1909	6378388		297.0		6356911.946	0.00672267
  5   WGS 72			6378135		298.26		6356750.52	0.006694318
  6   EVEREST			6377276.345	300.8017	6356075.413	0.006637847
  7   WGS 66			6378145		298.25		6356759.769	0.006694542
  8   GRS 1980			6378137		298.2572221	6356752.314	0.00669438
  9   AIRY			6377563.396	299.3249646	6356256.909	0.00667054
 10   MODIFIED EVEREST		6377304.063	300.8017	6356103.039	0.006637847
 11   MODIFIED AIRY		6377340.189	299.3249646	6356034.448	0.00667054
 12   WGS84			6378137.0	298.2572236	6356752.314	0.00669438
 13   SE ASIA						
      (MODIFIED FISCHER 1960)	6378155.0	298.3		6356773.32	0.006693422
 14   AUSTRALIAN NATIONAL	6378160.0	298.25		6356774.719	0.006694542
 15   KRASSOVSKY		6378245.0	298.3		6356863.019	0.006693422
 16   HOUGH			6378270.0	297.0		6356794.343	0.00672267
 17   MERCURY 1960						
      (FISCHER 1960)		6378166.0	298.3		6356784.284	0.006693422
 18   MODIFIED MERCURY 1968	6378150.0	298.3		6356768.337	0.006693422
 19   SPHERE			6370997.0	n/a		6370997.0	n/a
 20   BESSEL 1841(NAMIBIA)	6377483.865	299.1528128	6356165.383	0.006674372
 21   EVEREST 						
      (SABAH & SARAWAK)		6377298.556	300.802		6356097.571	0.00663784
 22   EVEREST (INDIA 1956)	6377301.243	300.8017	6356100.228	0.006637847
 23   EVEREST (MALAYSIA 1969)	6377295.664	300.8017	6356094.668	0.006637847
 24   EVEREST 						
      (MALAY. & SINGAPORE 1948)	6377304.063	300.8017	6356103.039	0.006637847
 25   EVEREST (PAKISTAN)	6377309.613	300.8017	6356108.571	0.006637847
 26   HAYFORD			6378388.0	297.0		6356911.946	0.00672267
 27   HELMERT 1906		6378200.0	298.3		6356818.17	0.006693422
 28   INDONESIAN 1974		6378160.000	298.247		6356774.504	0.006694609
 29   SOUTH AMERICAN 1969	6378160.000	298.25		6356774.719	0.006694542
 30   WGS 60			6378165.0	298.3		6356783.287	0.006693422
 31   HUGHES ELLIPSOID		6378273.0	298.279405	6356889.449	0.006693883
 32   TERASCAN SPHERE		6378145.0	n/a		6378145.0	n/a
 33   SGS 85			6378136.000	298.257		6356751.302	0.006694385
 34   GEM06			6378144.043	298.19516	6356754.883	0.006695771
 35   INTERNATIONAL		6378140.000	298.2529811	6356755.000	0.006694475

  -p	Option to specify the type of projection to interpret the associated
	  image with. If you need to do any geographic analysis with the data,
	  you will need to enter geolocation information as well.  This is
	  done via optional command line parameters.  All projections require
	  the following information:

	Projection type			(Described below)
	Projection parameters		(Described below)
        ULn				Upper Left Northing Coordinate
        ULe				Upper Left Easting Coordinate
        LRn				Lower Right Northing Coordinate
        LRe				Lower Right Easting Coordinate
        pdist				Projection Distance (units per pixel)

	NOTE: ULn, ULe, LRn, LRe, and pdist units are meters for all
		projections except GEOGRAPH in which case they are in degrees

	Projection |		Parameters			| Units
	-----------+--------------------------------------------+-----
	  GEOGRAPH | N/A					| N/A
	  PLSTEREO | Standard Parallel, Central Meridian	| degrees
	       UTM | Zone					| Zone number
	  LAMAZEQA | Central Latitude, Central Longitude,	| degrees
		   |	Radius of reference sphere		| meters
	    ALBERS | 1st Standard Parallel, 2nd Std Parallel,	| degrees
		   |	Central Meridian			| degrees


EXAMPLES:
     1)	If you want to make a ddr for a floating-point image named "bob.amp"
        with dimensions of 100 lines by 20 samples, type:
	
	makeddr bob 100 20 float
	
	And the program will create a 1-band DDR named "bob.ddr".

     2) If you want to make a ddr for a unsigned 2-byte int image named
	"mydem.img" with dimensions of 7480 lines by 5911 samples, 
	geolocations of Upper Left (6652899.87North, 440890.23East) and
  	Lower Right (6428499.87North, 618220.23East) in UTM Zone 8 at 
	30.0 meter pixel spacings, type:

	makeddr  -p UTM 8 6652899.87 440890.23 6428499.87 618220.23 30.0 \
	  mydem 7480 5911 short

	(don't type the '\' it is there only to signify a wrapped line)

	And the program will create a 1-band int*2 DDR named "mydem.ddr"
	with valid projection information.

     3) If you want to make a ddr for a 8 byte floating point image
        using a Geographic projection with dimensions 4096 lines by
	2000 samples, geolocations of Upper Left (65.5north -146.0east)
	and Lower Right (65.0north -145.0east), and 10.0 meter pixel
	spacing, type:
	
        makeddr -p GEOGRAPH 65.5 -146.0 65.0 -145.0 10.0 mydem 4096 2000 \
	  double
        
        (don't type the '\' it is there only to signify a wrapped line)

        And the program will create a 1-band real*8 DDR named "mydem.ddr"
        with valid projection information.

SEE ALSO:

HARDWARE/SOFTWARE LIMITATIONS:

BUGS:
	
AUTHORS: Orion Lawlor, Tom Logan, Patrick Denny

CONTACT:
	apd@asf.alaska.edu
	http://www.asf.alaska.edu/apd


trunk/src/makeddr/makeddr.c
/******************************************************************************
NAME:  makeddr -- creates a ddr file from the command line.

SYNOPSIS: 
   
   makeddr [-p <projection> <parameters> <ULn> <ULe> <LRn> <LRe> <pdist>]
   		[-d <datum code>] [-log <file>] <filename> <nl> <ns> <type> 

DESCRIPTION: 
	
 	Create a LAS 6.0 DDR (Data Descriptor Record) from the command line.
This is only useful when you don't have a DDR already (which is actually pretty
rare, but when you need one it's tough to get along without it).  This
might happen if you're importing data from an source that doesn't use
LAS (the Land Analysis System).

EXTERNAL ASSOCIATES:
    NAME:               USAGE:
    ---------------------------------------------------------------

FILE REFERENCES:
    NAME:               USAGE:
    ---------------------------------------------------------------

PROGRAM HISTORY:
    VERS:   DATE:  AUTHOR:      PURPOSE:
    ---------------------------------------------------------------
    1.0	    6/97   O. Lawlor    Needed DDR's for new data, to test
    				 processing programs before ingestion
    				 programs were finalized.
    2.0	    4/98   T. Logan     Need ability to create valid LAS ddr files
				 without using LAS. (*including geolocations*)
    3.0	    6/00   P. Denny	Added projections GEOGRAPH, PLSTEREO, 
				 LAMAZEQA, and ALBERS to the existing UTM
    3.1	    7/00   T. Logan     Modified Albers to store proj.parms in DMS
    3.11    7/01   R. Gens 	Added logfile switch
    3.5	    9/01   P. Denny	Fixed command line parsing
    				 added datum code option
    				 cleaned up the code
				 some error checking

HARDWARE/SOFTWARE LIMITATIONS:

ALGORITHM DESCRIPTION:

ALGORITHM REFERENCES:

BUGS:    

*****************************************************************************
*								            *
*   makeddr -- creates a ddr file from the command line.		    *
* Copyright (c) 2004, Geophysical Institute, University of Alaska Fairbanks   *
* All rights reserved.                                                        *
*                                                                             *
* Redistribution and use in source and binary forms, with or without          *
* modification, are permitted provided that the following conditions are met: *
*                                                                             *
*    * Redistributions of source code must retain the above copyright notice, *
*      this list of conditions and the following disclaimer.                  *
*    * Redistributions in binary form must reproduce the above copyright      *
*      notice, this list of conditions and the following disclaimer in the    *
*      documentation and/or other materials provided with the distribution.   *
*    * Neither the name of the Geophysical Institute nor the names of its     *
*      contributors may be used to endorse or promote products derived from   *
*      this software without specific prior written permission.               *
*                                                                             *
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" *
* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE   *
* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE  *
* ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE    *
* LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR         *
* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF        *
* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS    *
* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN     *
* CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)     *
* ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE  *
* POSSIBILITY OF SUCH DAMAGE.                                                 *
*                                                                             *
*       For more information contact us at:                                   *
*                                                                             *
*       Alaska Satellite Facility                                             *
*       Geophysical Institute                   http://www.asf.alaska.edu     *
*       University of Alaska Fairbanks          uso@asf.alaska.edu            *
*       P.O. Box 757320                                                       *
*       Fairbanks, AK 99775-7320                                              *
*                                                                             *
******************************************************************************/

#include "asf.h"
#include "ddr.h"
#include "las.h"
#include "locinc.h"
#include "cproj.h"

#define VERSION 3.5

enum projections {GEOGRAPH=1,PLSTEREO,UTM,LAMAZEQA,ALBERS};

void geograph(struct DDR*);
void plstereo(struct DDR*, char**);
void utm(struct DDR*, char**);
void lamazeqa(struct DDR*, char**);
void albers(struct DDR*, char**);
void corner_pdist(struct DDR*, char**);
void usage(char*);


int main(int argc, char **argv)
{
    struct DDR ddr;		/* DDR structure */
    char *key,datatype[7],proj[9];
    char fname[256];		/* DDR File name */
    char *projParam[4], *corners[5];/* pointers to projection information */
    int ii;			/* string size for projection */
    int nl,ns;			/* Number of lines, number of samples */
    int type=0;			/* data type value for DDR */
    int argNo=1;		/* command line argument counter*/
    int projection=0;		/* One of five projections, initialized as false */
    int datumFlag=0;		/* Did user specify datum code */

    logflag=0;			/* initialize flag for no logging */

/* allocate memory for pointers */
    for (ii=0; ii<4; ii++)
	projParam[ii] = (char*)MALLOC(sizeof(char)*128);
    for (ii=0; ii<5; ii++)
	corners[ii] = (char*)MALLOC(sizeof(char)*128);

/* initialize ddr structure */
    c_intddr(&ddr);

/* Parse command line arguments
------------------------------*/
/* Optional cla's */
    for (argNo=1; argNo<argc; ++argNo)
    {
	if (argc == argNo+4) break;
	if (argv[argNo][0] != '-') usage(argv[0]);
	switch (argv[argNo][1])
	{
	    case 'p': /*Get projection parameters & corners and meters per pixel (projParam & corners)*/
		strcpy(proj,argv[++argNo]);
		ii = strlen(proj);
		c_low2up(proj,&ii);
		if (0==strcmp(proj,"GEOGRAPH"))
		{
		    projection = GEOGRAPH;
		}
		else if (0==strcmp(proj,"PLSTEREO"))
		{
		    projection = PLSTEREO;
		    projParam[0] = argv[++argNo];
		    projParam[1] = argv[++argNo];
		}		    
		else if (0==strcmp(proj,"UTM"))
		{
		    projection = UTM;
		    projParam[0] = argv[++argNo];
		}
		else if (0==strcmp(proj,"LAMAZEQA"))
		{
		    projection = LAMAZEQA;
		    projParam[0] = argv[++argNo];
		    projParam[1] = argv[++argNo];
		    projParam[2] = argv[++argNo];
		}
		else if (0==strcmp(proj,"ALBERS"))
		{
		    projection = ALBERS;
		    projParam[0] = argv[++argNo];
		    projParam[1] = argv[++argNo];
		    projParam[2] = argv[++argNo];
		    projParam[3] = argv[++argNo];
		}
		else 
		{
		    sprintf(errbuf, "ERROR: %s does not support the %s projection\n\n",argv[0],proj);
		    printErr(errbuf);
		}
		corners[0] = argv[++argNo];/*uln*/
		corners[1] = argv[++argNo];/*ule*/
		corners[2] = argv[++argNo];/*lrn*/
		corners[3] = argv[++argNo];/*lre*/
		corners[4] = argv[++argNo];/*meters per pixel*/
		break;
	    case 'd': /*datum code*/
		ddr.datum_code = atoi(argv[++argNo]);
		ddr.valid[2] = VALID;
		datumFlag=1;
		break;
	    case 'l': /*use a logfile*/
		sprintf(logFile,"%s",argv[++argNo]);
		fLog = FOPEN(logFile, "a");
		logflag=1;
		break;
	    default: /*invalid arg*/
		usage(argv[0]);
		break;
	}
    }
/* Required cla's */
    if (argc == argNo+4)
    {
	strcpy(fname,argv[argNo++]);
	nl=atol(argv[argNo++]);
	ns=atol(argv[argNo++]);
	key=argv[argNo++];
	if ((0==strcmp(key,"byte")) || (0==strcmp(key,"char"))) {type=1; strcpy(datatype,"byte");}
	else if (0==strcmp(key,"short")) {type=2; strcpy(datatype,"short");}
	else if (0==strcmp(key,"long")) {type=3; strcpy(datatype,"long");}
	else if (0==strcmp(key,"float")) {type=4; strcpy(datatype,"float");}
	else if (0==strcmp(key,"double")) {type=5; strcpy(datatype,"double");}
	else {sprintf(errbuf, "ERROR: Unrecognized data type '%s'.\n\n",key); printErr(errbuf);}
    }
    else usage(argv[0]);
    
/* Finally, we can begin DDR creation 
-----------------------------------*/
    printf("\nCreating ddr...\n");
    printf("    Lines:     %i\n", nl);
    printf("    Samples:   %i\n", ns);
    printf("    Data type: %s\n", datatype);
 
/* Enter default info into DDR structure
----------------------------------------*/
    ddr.nl = nl;
    ddr.ns = ns;
    ddr.nbands = (int)1;
    ddr.dtype = (int)type;

    ddr.pdist_x = 12.5;			/* dummy values */
    ddr.pdist_y = 12.5;

    if (!datumFlag)
    {
	ddr.datum_code = (int)0;
	ddr.valid[2] = INVAL;		/* Datum Code */
    }

    strcpy(ddr.proj_units,"METERS");	/* Changed to degrees if GEOGRAPH projection */
    ddr.valid[4] = INVAL;		/* Projection Units (can be modified by projection fuctions */

    ddr.valid[5] = INVAL;		/* Projection Distance (can be modified by *corner_pdist()) */

    ddr.line_inc = 1.0;
    ddr.sample_inc = 1.0;
    ddr.valid[7] = VALID;		/* Increment */

    ddr.master_line = (int)1;
    ddr.master_sample = (int)1;

/* Enter projection specific info into DDR structure
---------------------------------------------------*/
    if (projection)
    { 
	if (projection == GEOGRAPH)	 geograph(&ddr);
	else if (projection == PLSTEREO) plstereo(&ddr, projParam);
	else if (projection == UTM) 	 utm(&ddr, projParam);
	else if (projection == LAMAZEQA) lamazeqa(&ddr, projParam);
	else if (projection == ALBERS)	 albers(&ddr, projParam);
	else {sprintf(errbuf,"\tERROR: %s does not support the %s projection\n\n",argv[0],proj); printErr(errbuf);}
	corner_pdist(&ddr, corners);
    }

/* Write to error log if specified
---------------------------------*/
    if (logflag)
    {
	StartWatchLog(fLog);
	printLog("Creating ddr...\n");
    	sprintf(logbuf,"    Lines:     %i\n", nl);	   printLog(logbuf);
    	sprintf(logbuf,"    Samples:   %i\n", ns);	   printLog(logbuf);
	sprintf(logbuf,"    Data type: %s\n\n", datatype); printLog(logbuf);
    }

/* Put DDR structure info into specified file
---------------------------------------------*/
    if (c_putddr(fname, &ddr) != E_SUCC)
       {sprintf(errbuf,"\tERROR: Error returned from putddr\n\n"); printErr(errbuf);}
    printf("\nSuccessfully created %s DDR file\n\n",fname);
    return(0);

} /******** END PROGRAM ********/




/* Enter DDR info for geographic
-------------------------------*/
void geograph(struct DDR *ddr)
{
	printf("\n    Adding GEOGRAPH projection information...\n");

	ddr->proj_code = (int)0;
	ddr->valid[0] = VALID;		/* Projection Code	*/
	ddr->zone_code = (int)62;
	ddr->valid[1] = VALID;		/* Zone Code N/A	*/
	ddr->valid[3] = VALID;	        /* Projection Parameters*/
	strcpy(ddr->proj_units,"DEGREES");
	ddr->valid[4] = VALID;		/*  Projection Units	*/
}

/* Enter DDR info for Polar Stereographic
----------------------------------------*/
void plstereo(struct DDR *ddr, char **projParam)
{
    	double par_deg, lon_deg, par_dms, lon_dms;
	
    	par_deg = atof(projParam[0]);
    	lon_deg = atof(projParam[1]);

	printf("\n    Adding PLSTEREO projection information...\n");

	printf("\tStandard Parallel: %f degrees\n",par_deg);
    	if (E_SUCC != c_degdms(&par_deg,&par_dms,"DEG","LAT"))
	  {sprintf(errbuf,"ERROR: %f is not a valid latitude\n\n",par_deg); printErr(errbuf);}
	printf("\tCentral Longitude: %f degrees\n",lon_deg);
	if (E_SUCC != c_degdms(&lon_deg,&lon_dms,"DEG","LON"))
	  {sprintf(errbuf,"ERROR: %f is not a valid longitude\n\n",lon_deg); printErr(errbuf);}

	ddr->proj_code = (int)6;
	ddr->valid[0] = VALID;		/* Projection Code	*/
	ddr->zone_code = (int)62;
	ddr->valid[1] = VALID;		/* Zone Code N/A	*/
	ddr->valid[3] = VALID;	        /* Projection Parameters*/
	ddr->valid[4] = VALID;		/* Projection Units	*/

	ddr->proj_coef[0] = (double)6378273;
	ddr->proj_coef[1] = (double)0.006693883;
	ddr->proj_coef[5] = par_dms;
	ddr->proj_coef[4] = lon_dms;
}

/* Enter DDR info for Universal Transverse Mercator
--------------------------------------------------*/
void utm(struct DDR *ddr, char **projParam)
{
	int zone = atoi(projParam[0]);

	printf("\n    Adding UTM projection information...\n");

    	printf("\tZone: %i\n",zone);

	ddr->proj_code = (int)1;
        ddr->valid[0] = VALID;		/* Projection Code       */
	ddr->zone_code = (int)zone;
        ddr->valid[1] = VALID;		/* Zone Code 	         */
        ddr->valid[3] = VALID;	        /* Projection Parameters */
	ddr->valid[4] = VALID;		/* Projection Units	*/
}

/* Enter DDR info for Lambert Azimuthal Equal Area
-------------------------------------------------*/
void lamazeqa(struct DDR *ddr, char **projParam)
{
    	double lat_deg, lon_deg, radius, lat_dms, lon_dms;

    	lat_deg = atof(projParam[0]);
    	lon_deg = atof(projParam[1]);
    	radius  = atof(projParam[2]);

	printf("\n    Adding LAMAZEQA projection information...\n");

	printf("\tCentral Latitude: %f degrees\n",lat_deg);
	if (E_SUCC != c_degdms(&lat_deg,&lat_dms,"DEG","LAT"))
	  {sprintf(errbuf,"ERROR: %f is not a valid latitude\n\n",lat_deg); printErr(errbuf);}
	printf("\tCentral Longitude: %f degrees\n",lon_deg);
	if (E_SUCC != c_degdms(&lon_deg,&lon_dms,"DEG","LON"))
	  {sprintf(errbuf,"ERROR: %f is not a valid longitude\n\n",lon_deg); printErr(errbuf);}

	printf("\tRadius of reference sphere: %f meters\n", radius);

	ddr->proj_code = (int)11;
        ddr->valid[0] = VALID;		/* Projection Code       */
	ddr->zone_code = (int)62;
        ddr->valid[1] = VALID;		/* Zone Code N/A    	 */
        ddr->valid[3] = VALID;	        /* Projection Parameters */
	ddr->valid[4] = VALID;		/* Projection Units	*/
	
    	ddr->proj_coef[0] = radius;
    	ddr->proj_coef[4] = lat_dms;
    	ddr->proj_coef[5] = lon_dms;
}

/* Enter DDR info for Albers Conical Equal Area
----------------------------------------------*/
void albers(struct DDR *ddr, char **projParam)
{
    	double lat1, lat2, cenMer, latorigin;

    	lat1 = atof(projParam[0]);
    	lat2 = atof(projParam[1]);
    	cenMer = atof(projParam[2]);
	latorigin = atof(projParam[3]);

	/* Report parameters and put in .ddr in dms form */
	printf("\n    Adding ALBERS projection information...\n");

    	printf("\tFirst Standard Parallel Parameter: %f degrees\n", lat1);
	if (E_SUCC != c_degdms(&lat1,&(ddr->proj_coef[2]),"DEG","LAT"))
	  {sprintf(errbuf,"ERROR: %f is not a valid latitude\n\n",lat1); printErr(errbuf);}
    	printf("\tSecond Standard Parallel Parameter: %f degrees\n", lat2);
	if (E_SUCC != c_degdms(&lat2,&(ddr->proj_coef[3]),"DEG","LAT"))
	  {sprintf(errbuf,"ERROR: %f is not a valid latitude\n\n",lat2); printErr(errbuf);}
    	printf("\tCentral Meridian: %f degrees\n", cenMer);
	if (E_SUCC != c_degdms(&cenMer,&(ddr->proj_coef[4]),"DEG","LON"))
	  {sprintf(errbuf,"ERROR: %f is not a valid longitude\n\n",cenMer); printErr(errbuf);}
	printf("\tLatitude of Origin: %f degrees\n", latorigin);
        if (E_SUCC != c_degdms(&latorigin,&(ddr->proj_coef[5]),"DEG","LAT"))
	  {sprintf(errbuf,"ERROR: %f is not a valid latitude\n\n",latorigin); printErr(errbuf);}

	ddr->proj_code = (int)3;
	ddr->valid[0] = VALID;		/* Projection Code       */
	ddr->zone_code = (int)62;
	ddr->valid[1] = VALID;		/* Zone Code N/A    	 */
	ddr->valid[3] = VALID;	        /* Projection Parameters */
	ddr->valid[4] = VALID;		/* Projection Units	*/
}

/* Enter corner coordinates and pixel size to DDR struct
   necessary for all projections
--------------------------------------------------------*/
void corner_pdist(struct DDR *ddr, char **corners)
{
	double uln,ule,lrn,lre;
	double pdist;

	uln = atof(corners[0]);
	ule = atof(corners[1]);
        lrn = atof(corners[2]);
	lre = atof(corners[3]);
	pdist = atof(corners[4]);

    	printf("\n    Pixel Size: %f meters\n", pdist);

	ddr->upleft[0]  = uln; ddr->upleft[1] = ule;
	ddr->loleft[0]  = lrn; ddr->loleft[1] = ule;
	ddr->upright[0] = uln; ddr->upright[1] = lre;
	ddr->loright[0] = lrn; ddr->loright[1] = lre;
        ddr->valid[6] = VALID;		/* Corner Coordinates    */

	ddr->pdist_x = ddr->pdist_y = pdist;
        ddr->valid[5] = VALID;		/* Projection Distance   */
}

void usage(char *name)
{
    printf("\nUSAGE:\n");
    printf("  %s [-p <projection> <projection parameters> <uln> <ule> <lrn> <lre> <pdist>]\n",name);
    printf("\t [-d <datum code>] [-log <file>] <filename> <nl> <ns> <type>\n\n");
    printf("  Options:\n");
    printf("\t-l   - Option to have output written to a log <file>\n");
    printf("\t-d   - Option to enter the <datum code>\n");
    printf("\t	 	(see man page for more information)\n");
    printf("\t-p   - Specific projection type & parameters\n");
    printf("\t		(see man page for more information)\n");
    printf("\t\t   Projections	| Parameters\n");
    printf("\t\t   -------------+-----------------------------------------\n");
    printf("\t\t   GEOGRAPH	| N/A\n");
    printf("\t\t   PLSTEREO	| Standard Parallel, Central Meridian\n");
    printf("\t\t   UTM		| Zone\n");
    printf("\t\t   LAMAZEQA	| Central Latitude, Central Longitude,\n");
    printf("\t\t		|	Radius of reference sphere\n");
    printf("\t\t   ALBERS	| 1st Standard Parallel, 2nd Std Parallel,\n");
    printf("\t\t		|	Central Meridian\n");
    printf("  Required inputs:\n");
    printf("\tfilename - Base name for .ddr (use same name as the base image name)\n");
    printf("\tnl\t - Number of lines in image\n");
    printf("\tns\t - Number of samples in image\n");
    printf("\ttype\t - Data type used for image.  Types are:\n");
    printf("\t\t\tbyte, short, long, float, double\n\n");
    printf("  Description:\n");
    printf("\tThe makeddr program will construct a 1-band ddr\n");
    printf("\tfile with nl lines, ns samples, and data type.\n\n");
    printf("  Version %.2f, ASF SAR Tools\n",VERSION);
    exit(1);
}

